elk/run.py

```diff
@@ -0,0 +1,111 @@
+import csv
+import os
+import random
+from abc import ABC
+from dataclasses import dataclass, field
+from functools import partial
+from pathlib import Path
+from typing import TYPE_CHECKING, List, Optional, Union, cast
+
+import numpy as np
+import torch
+import torch.multiprocessing as mp
+from torch import Tensor
+from tqdm import tqdm
+
+from datasets import DatasetDict, Split
+from elk.extraction.extraction import extract
+from elk.files import create_output_directory, save_config, save_meta
+from elk.logging import save_debug_log
+from elk.training.preprocessing import normalize
+from elk.utils.data_utils import get_layers, select_train_val_splits
+from elk.utils.gpu_utils import select_usable_devices
+from elk.utils.typing import assert_type, upcast_hiddens
+
+if TYPE_CHECKING:
+    from elk.evaluation.evaluate import Eval
+    from elk.training.train import Elicit
+
+
+@dataclass
+class Run(ABC):
+    cfg: Union["Elicit", "Eval"]
+    eval_headers: List[str]
+    out_dir: Optional[Path] = None
+    dataset: DatasetDict = field(init=False)
+
+    def __post_init__(self):
+        # Extract the hidden states first if necessary
+        self.dataset = extract(self.cfg.data, max_gpus=self.cfg.max_gpus)
+
+        self.out_dir = create_output_directory(self.out_dir)
+        save_config(self.cfg, self.out_dir)
+        save_meta(self.dataset, self.out_dir)
+
+    def make_reproducible(self, seed: int):
+        """Make the run reproducible by setting the random seed."""
+
+        np.random.seed(seed)
+        random.seed(seed)
+        torch.manual_seed(seed)
+
+    def get_device(self, devices, world_size: int) -> str:
+        """Get the device for the current process."""
+
+        rank = os.getpid() % world_size
+        device = devices[rank]
+        return device
+
+    def prepare_data(
+        self,
+        device: str,
+        layer: int,
+        priorities: dict = {Split.TRAIN: 0, Split.VALIDATION: 1, Split.TEST: 2},
+    ) -> tuple:
+        """Prepare the data for training and validation."""
+
+        with self.dataset.formatted_as("torch", device=device, dtype=torch.int16):
+            train_split, val_split = select_train_val_splits(
+                self.dataset, priorities=priorities
+            )
+            train, val = self.dataset[train_split], self.dataset[val_split]
+
+            train_labels = cast(Tensor, train["label"])
```

Comment:
We should also use `assert_type` here. Basically the rule is to never use `cast`.

Internally `assert_type` does call `cast`, but it also does a runtime check that the type is in fact what you're saying it is, while `cast` just takes your word for it.